external/llvm-project/mlir/lib/Dialect/LLVMIR/Transforms/SoftwareBf16.cpp

```diff
@@ -0,0 +1,364 @@
+    Type destType = op.getRes().getType();
+    Type srcElemType = srcType;
+    if (auto shaped = srcType.dyn_cast<ShapedType>())
+      srcElemType = shaped.getElementType();
+
+    Type f32 = rewriter.getF32Type();
+    if (srcElemType != f32)
+      return failure();
+
+    Type bitcastType = rewriter.getI32Type();
+    if (auto srcShaped = srcType.dyn_cast<ShapedType>())
+      bitcastType = srcShaped.clone(bitcastType);
+
+    Type i16 = rewriter.getIntegerType(16);
+    if (auto destShaped = destType.dyn_cast<ShapedType>()) {
+      if (destShaped.getElementType() != i16)
+        return failure();
+    } else if (destType != i16)
+      return failure();
+
+    // a = bitcast f32 value to i32
+    // b = (a + 32767) >> 16
+    // c = ((a >> 16) & 1)
+    // d = b + c
+    // truncate (d << 16) to i16 and return this i16
+    Value bitcastop =
+        rewriter.create<LLVM::BitcastOp>(loc, bitcastType, op.getArg());
+
+    Value constantSixteen = getLlvmI32Const(loc, rewriter, bitcastType, 16);
+    Value shiftValue = rewriter.create<LLVM::LShrOp>(
+        loc, bitcastType, bitcastop, constantSixteen);
+
+    Value constantOne = getLlvmI32Const(loc, rewriter, bitcastType, 1);
+    Value andValue = rewriter.create<LLVM::AndOp>(loc, shiftValue, constantOne);
+
+    Value constantBig = getLlvmI32Const(loc, rewriter, bitcastType, 32767);
+    Value addBigValue =
+        rewriter.create<LLVM::AddOp>(loc, bitcastop, constantBig);
+    Value shiftBigValue = rewriter.create<LLVM::LShrOp>(
+        loc, bitcastType, addBigValue, constantSixteen);
+
+    Value addValue = rewriter.create<LLVM::AddOp>(loc, andValue, shiftBigValue);
+
+    Value shiftBeforeTruncValue = rewriter.create<LLVM::LShrOp>(
+        loc, bitcastType, addValue, constantSixteen);
+    Value truncValue =
+        rewriter.create<LLVM::TruncOp>(loc, destType, shiftBeforeTruncValue);
+    rewriter.replaceOp(op.getOperation(), {truncValue});
+
+    return success();
+  }
+};
+
+} // namespace
+
+static void replaceBF16WithI16(Operation *op, TypeConverter &converter) {
+  if (auto func = dyn_cast<LLVM::LLVMFuncOp>(op)) {
+    auto funcType = func.getFunctionType();
+    func.setType(converter.convertType(funcType));
+    for (auto arg : func.getArguments())
+      arg.setType(converter.convertType(arg.getType()));
+  } else if (auto globalOp = dyn_cast<LLVM::GlobalOp>(op)) {
+    auto globalType = globalOp.getType();
+    globalOp.setGlobalTypeAttr(
+        mlir::TypeAttr::get(converter.convertType(globalType)));
+  } else {
+    for (unsigned idx = 0; idx < op->getNumOperands(); idx++) {
+      auto type = converter.convertType(op->getOperand(idx).getType());
+      op->getOperand(idx).setType(type);
+    }
+    for (unsigned idx = 0; idx < op->getNumResults(); idx++) {
+      auto type = converter.convertType(op->getResult(idx).getType());
+      op->getResult(idx).setType(type);
+    }
+  }
+  return;
+}
+
+static void populateSoftwareBF16Patterns(LLVMTypeConverter &converter,
```

Comment:
I insist this can just be a `TypeConverter` - do you have counterexamples?

Reply:
(you _might_ need to add the fallback rule that types convert to themselves if not otherwise specified ... I can't quite remember how the type conversion machinery handles the default case)

Reply:
LLVMTypeConverter can handle the default case, while TypeConverter cannot

Reply:
LLVMTypeConverter doesn't actually handle the default case - it just has a bunch of conversions. You can get much the same behavior by having the first conversion you add be `(Type t) { return t; }`

Reply:
... sorry, didn't see your commit